File: vault-ios/VaultMobile/MemorySecureStorage.swift
import Foundation

public class MemorySecureStorage: SecureStorage {
    private var data: [String: String]

    public init() {
        self.data = [String: String]()
    }

    public func getData() -> [String: String] {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        return data
    }

    public func getItem(key: String) throws -> String? {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        return data[key]
    }

    public func setItem(key: String, value: String) throws {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        data[key] = value
    }

    public func removeItem(key: String) throws {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        data.removeValue(forKey: key)
    }

    public func clear() throws {
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }

        data.removeAll()
    }
}

File: vault-core/src/secure_storage/memory_secure_storage.rs
use std::{
    collections::HashMap,
    sync::{Arc, RwLock, RwLockReadGuard},
};

use super::SecureStorage;

#[derive(Debug)]
pub struct MemorySecureStorage {
    data: Arc<RwLock<HashMap<String, String>>>,
}

impl MemorySecureStorage {
    pub fn new() -> Self {
        Self::new_with_data(HashMap::new())
    }

    pub fn new_with_data(data: HashMap<String, String>) -> Self {
        Self {
            data: Arc::new(RwLock::new(data)),
        }
    }

    pub fn get_data(&self) -> RwLockReadGuard<'_, HashMap<String, String>> {
        self.data.read().unwrap()
    }
}

impl SecureStorage for MemorySecureStorage {
    fn get_item(&self, key: &str) -> Result<Option<String>, String> {
        Ok(self.data.read().unwrap().get(key).cloned())
    }

    fn set_item(&self, key: &str, value: &str) -> Result<(), String> {
        self.data
            .write()
            .unwrap()
            .insert(key.to_owned(), value.to_owned());

        Ok(())
    }

    fn remove_item(&self, key: &str) -> Result<(), String> {
        self.data.write().unwrap().remove(key);

        Ok(())
    }

    fn clear(&self) -> Result<(), String> {
        self.data.write().unwrap().clear();

        Ok(())
    }
}
